smartsim/_core/utils/helpers.py

```diff
@@ -302,3 +306,43 @@ def decode_cmd(encoded_cmd: str) -> t.List[str]:
     cleaned_cmd = decoded_cmd.decode("ascii").split("|")
 
     return cleaned_cmd
+
+
+@t.final
+@dataclass(frozen=True)
+class _Pipeline(t.Generic[_T]):
```

Comment:
The builder pattern normally returns the original object instead of creating a new _Pipeline each time. One bonus is that you have an actual description of the expected operations and potentially lazy execution.

Reply:
I was actually trying to explicitly avoid the builder pattern and was going for something similar to bash's `|` or OCaml's `|>`to keep everything "immutable" (you can see that I do end up using this with fns that have side effects, hahaha).

We can still do lazy execution, we just need to move where the fn gets called to the `get_value` method instead
```py
class _Pipeline[T]:
    val: Callable[[], T]

    @classmethod
    def from(val: T) -> _Pipeline[T]:
        return cls(lambda: val)

    def then(self, fn: Callable[U, V]) -> _Pipeline[U, V]:
        return _Pipeline(lambda: fn(self.get_value()))

    def get_value(self) -> T:
        return self.val()

some_data = ...
p = (_Pipeline.from(some_data)
     .then(these)
     .then(methods)
     .then(wont)
     .then(be)
     .then(evaluated))

# until now
p.get_value()
```

Reply:
But again, totally willing to scrap this whole data structure if we agree that it is unreadable!

Reply:
`_Pipeline` has been removed with https://github.com/CrayLabs/SmartSim/pull/470/commits/97292df741aba1acb644f39312abfb44380cb556 🎉